### PR TITLE
Transport error handling

### PR DESCRIPTION
**Pull request by Scooletz**
Connects to https://github.com/Particular/PlatformDevelopment/issues/832

This PR beside the alignment to the new recoverability extracts two implementations of `ReceiveStrategy`: `AtLeastOnce` and `AtMostOnce`. This clarifies a lot a different set of responsibilities which they have, when dispatching a message.

Notes:
- currently ASQ Transport uses only the `main serializer` to serialize/deserialize `MessageWrapper` and this cannot be changed. Because of this, one of the serializer tests is `Ignored` for now.
- an internal `SettingsHolder.Merge(...)` method is called via reflection to enable combining serializer settings with default settings (a'la `SerializationFeature`).

### CODE REVIEW

**Comment by Scooletz** [comment]
@SeanFeldman could you take a look at the current version. I've cleaned it up to align to our discussion/understanding of the context. /cc @yvesgoeleven

**Comment by SeanFeldman** [comment]
:+1: @Scooletz looks good. I'll also raise a question about the case of `AtMostOnceReceiveStrategy`, when we receive and delete. What sort of expectation core is making in that scenario.

**Comment by Scooletz** [comment]
I know this needs a rebase but still ping @Particular/azure-storage-queues-maintainers to review again.

File: src/Transport/ReceiveStrategy.cs [hunk 1]
```diff
@@ -120,22 +109,27 @@ public AtLeastOnceReceiveStrategy(Func<MessageContext, Task> pipeline, Func<Erro
                             }
                         }
                     }
+                    catch (LeaseTimeoutException)
+                    {
+                        // The lease has expired and cannot be used any longer to Ack or Nack the message.
+                        throw;
```
Comment (SeanFeldman) [inline src/Transport/ReceiveStrategy.cs]: Why to catch if we throw anyways?

**Comment by SeanFeldman** [comment]
Not sure I like the fact hat #126 is mixed in with the new recoverability seam.

File: src/Transport/ReceiveStrategy.cs (hunk 1 above)
Comment (Scooletz) [inline src/Transport/ReceiveStrategy.cs]: As this is a specific exception dealing with the fact that the lease / popReceipt expired.

Comment (SeanFeldman) [inline src/Transport/ReceiveStrategy.cs]: If you mean this is for readability/documentation purpose, perhaps song in
the content issue URL?

**Comment by Scooletz** [comment]
I don't like it as well, but only in the new recoverability the error was visible. Additionally, it influenced and showed a few cases to help me & @tmasternak clarify the approach for ASQ.
I'll separate this in n separate PR and `push force` the recent changes.

File: src/Transport/ReceiveStrategy.cs (hunk 1 above)
Comment (Scooletz) [inline src/Transport/ReceiveStrategy.cs]: I'm not sure I'm following. Let me rephrase. This is the explicit exception thrown on the `popReceipt` expiration. This needs to be handled separately as it changes the flow. We know that we are not capable of `Nack`/`Ack` hence there's no way of passing it to the `errorPipeline` (we won't be able to react properly).

**Comment by Scooletz** [comment]
#126 subtracted from this PR. All squashed. Ping @SeanFeldman /cc @tmasternak

**Comment by Scooletz** [comment]
Ping @Particular/azure-storage-queues-maintainers for a review.

**Comment by Scooletz** [comment]
Ping @SeanFeldman 
This needs a post merge review. It has been merged as a enabler for the rest of the packages to be built.

File: src/Transport/MessagePump.cs [hunk 2]
```diff
@@ -156,6 +142,7 @@ static int EstimateDegreeOfReceiveParallelism(int maxConcurrency)
                 }
                 catch (Exception ex)
                 {
+                    // TODO: Should it be the circuit breaker or the onCriticalError call?
```
Comment (SeanFeldman) [inline src/Transport/MessagePump.cs]: I'd think yes. The code is already doing this. Is there still a need in this TODO?

File: src/Transport/MessageRetrieved.cs [hunk 3]
```diff
@@ -38,34 +39,14 @@ public MessageWrapper Unwrap()
...
-                await azureQueue.DeleteMessageAsync(rawMessage).ConfigureAwait(false);
-            }
-            catch (StorageException ex)
-            {
-                if (ex.RequestInformation.HttpStatusCode != 404)
-                {
-                    throw;
-                }
-            }
+            await azureQueue.DeleteMessageAsync(rawMessage).ConfigureAwait(false);
```
Comment (SeanFeldman) [inline src/Transport/MessageRetrieved.cs]: Just return `Task` (`return azureQueue.DeleteMessageAsync(rawMessage)`)?

File: src/Transport/ReceiveStrategy.cs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            return context;
+        }
+
+        static ILog Logger = LogManager.GetLogger(typeof(ReceiveStrategy));
+
+        /// <summary>
+        /// At-most-once receive strategy receives at most once, acking first, then processing the message.
+        /// If the pipeline fails, the message is not processed any longer. No first or second level retries are executed.
+        /// </summary>
+        sealed class AtMostOnceReceiveStrategy : ReceiveStrategy
```
Comment (SeanFeldman) [inline src/Transport/ReceiveStrategy.cs]: Split it from the `ReceiveStrategy` to make GH searching easy.

File: src/Transport/ReceiveStrategy.cs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+                    await errorPipe(context).ConfigureAwait(false);
+                }
+            }
+
+            readonly Func<MessageContext, Task> pipeline;
+            readonly Func<ErrorContext, Task<ErrorHandleResult>> errorPipe;
+            static readonly CancellationTokenSource EmptyCancellationTokenSource = new CancellationTokenSource();
+        }
+
+        sealed class AtLeastOnceReceiveStrategy : ReceiveStrategy
```
Comment (SeanFeldman) [inline src/Transport/ReceiveStrategy.cs]: Split it from the `ReceiveStrategy` to make GH searching easy?

**Comment by SeanFeldman** [comment]
@Scooletz GH is failing to render the diff, too many files (316)
Not sure how to review properly. Did what could :)

File: src/Transport/MessagePump.cs (hunk 2 above)
Comment (Scooletz) [inline src/Transport/MessagePump.cs]: Nope, I'll clean this up.
